lib/Dialect/Polynomial/Conversions/PolynomialToModArith/PolynomialToModArith.cpp

```diff
@@ -0,0 +1,456 @@
+            [&](auto op) { inputType = op.getOperand().getType(); })
+        .template Case<AddOp, SubOp, MulScalarOp>(
+            [&](auto op) { inputType = op.getOperands()[0].getType(); });
+    if (inputType) {
+      inputTypeConverted = typeConverter->convertType(inputType);
+      inputTensorType = cast<RankedTensorType>(inputTypeConverted);
+      inputShape = inputTensorType.getShape()[0];
+      inputElementType = inputTensorType.getElementType();
+      inputElementWidth = inputElementType.getIntOrFloatBitWidth();
+      inputElementIntegerType = dyn_cast<IntegerType>(inputElementType);
+      inputElementFloatType = dyn_cast<FloatType>(inputElementType);
+    }
+  }
+
+  // polynomial type related
+  PolynomialType polynomialType;
+
+  RingAttr ringAttr;
+
+  Type coefficientType;
+  unsigned coefficientTypeWidth;
+  IntegerType coefficientIntegerType;
+  FloatType coefficientFloatType;
+
+  IntegerAttr coefficientModulus;
+  IntPolynomialAttr polynomialModulus;
+
+  APInt coefficientNaturalModulus;
+  APInt coefficientModulusValue;
+  bool coefficientModulusIsNotNatural;
+  unsigned coefficientModulusWidth;
+
+  // tensor related
+  Type resultType;
+  Type resultTypeConverted;
+  RankedTensorType resultTensorType;
+  int64_t resultShape;
+  Type resultElementType;
+  unsigned resultElementWidth;
+  IntegerType resultElementIntegerType;
+  FloatType resultElementFloatType;
+
+  Type inputType;
+  Type inputTypeConverted;
+  RankedTensorType inputTensorType;
+  int64_t inputShape;
+  Type inputElementType;
+  unsigned inputElementWidth;
+  IntegerType inputElementIntegerType;
+  FloatType inputElementFloatType;
+};
+
+template <typename Op>
+Value modArithReduceOp(ConvertCommon<Op> &c, ImplicitLocOpBuilder &b,
+                       ConversionPatternRewriter &rewriter, Value v,
+                       int64_t shape = 0) {
+  // why not c.resultTensorType here
+  // special case for inputShape != resultShape for FromTensor
+  auto resultTensorType = RankedTensorType::get(shape, c.coefficientType);
+  Type resultType = shape ? Type(resultTensorType) : Type(c.coefficientType);
+
+  // Extend for mod_arith::reduceOp in case modulus as large as underlying
+  // type
+  if (c.resultElementWidth == c.coefficientModulusWidth) {
+    auto reduceIntegerType =
+        rewriter.getIntegerType(c.coefficientModulusWidth + 1);
+    auto reduceTensorType = RankedTensorType::get(shape, reduceIntegerType);
+    Type reduceType = shape ? Type(reduceTensorType) : Type(reduceIntegerType);
+
+    v = b.create<arith::ExtSIOp>(reduceType, v);
+    v = b.create<mod_arith::ReduceOp>(reduceType, v, c.coefficientModulus);
+    v = b.create<arith::TruncIOp>(resultType, v);
+  } else if (c.coefficientModulusWidth != 0 &&
+             c.resultElementWidth > c.coefficientModulusWidth) {
+    v = b.create<mod_arith::ReduceOp>(resultType, v, c.coefficientModulus);
+  }
+  // else float or natural power of two modulus larger than underlying type.
```

Comment:
Why don't we need the op if it is a float? Should we really reduce it but can't because it doesn't support floats at the moment?

Reply:
The polynomial dialect does not have definition for `coefficientModulus` when `coeffcientType` is a float

```
if the coefficient type is integral, whose coefficients are taken modulo some statically known modulus (coefficientModulus).
```